File: src/interface/map/oocMap.c

#include "oocMap.h"
#include "oocMap.r"
#include "oocInterface.h"
#include "oocObject.h"

static OOC_MapClass* MapClass;
static OOC_MapClass MapClassInstance;

static void* ooc_mapClassInit(void) {
    if (ooc_classNew(&MapClassInstance,
                    "Map",
                    (size_t)0,
                    sizeof(const OOC_MapClass),
                    ooc_interfaceClass(),
                    OOC_MODIFIER_ABSTRACT,
                    0) != OOC_ERROR_NONE) {
        return NULL;
    }
    return &MapClassInstance;
}

void* ooc_mapClass() {
    if (!MapClass) {
        MapClass = ooc_mapClassInit();
    }
    return MapClass;
}

size_t ooc_mapSize(void* self) {
    if (!self) {
        return 0;
    }
    const OOC_MapVtable* vtable = ooc_getInterfaceVtable(self, ooc_mapClass());
    if (!vtable || !vtable->size) {
        return 0;
    }
    return vtable->size(self);
}

bool ooc_mapIsEmpty(void* self) {
    if (!self) {
        return true;
    }
    const OOC_MapVtable* vtable = ooc_getInterfaceVtable(self, ooc_mapClass());
    if (!vtable || !vtable->isEmpty) {
        return true;
    }
    return vtable->isEmpty(self);
}

bool ooc_mapContainsKey(void* self, void* key) {
    if (!self) {
        return false;
    }
    const OOC_MapVtable* vtable = ooc_getInterfaceVtable(self, ooc_mapClass());
    if (!vtable || !vtable->containsKey) {
        return false;
    }
    return vtable->containsKey(self, key);
}

bool ooc_mapContainsValue(void* self, void* value) {
    if (!self) {
        return false;
    }
    const OOC_MapVtable* vtable = ooc_getInterfaceVtable(self, ooc_mapClass());
    if (!vtable || !vtable->containsValue) {
        return false;
    }
    return vtable->containsValue(self, value);
}

void* ooc_mapGet(void* self, void* key) {
    if (!self) {
        return NULL;
    }
    const OOC_MapVtable* vtable = ooc_getInterfaceVtable(self, ooc_mapClass());
    if (!vtable || !vtable->get) {
        return NULL;
    }
    return vtable->get(self, key);
}

OOC_Error ooc_mapPut(void* self, void* key, void* value) {
    if (!self) {
        return OOC_ERROR_INVALID_ARGUMENT;
    }
    const OOC_MapVtable* vtable = ooc_getInterfaceVtable(self, ooc_mapClass());
    if (!vtable || !vtable->put) {
        return OOC_ERROR_NOT_IMPLEMENTED;
    }
    return vtable->put(self, key, value);
}

OOC_Error ooc_mapRemove(void* self, void* key) {
    if (!self) {
        return OOC_ERROR_INVALID_ARGUMENT;
    }
    const OOC_MapVtable* vtable = ooc_getInterfaceVtable(self, ooc_mapClass());
    if (!vtable || !vtable->remove) {
        return OOC_ERROR_NOT_IMPLEMENTED;
    }
    return vtable->remove(self, key);
}

OOC_Error ooc_mapClear(void* self) {
    if (!self) {
        return OOC_ERROR_INVALID_ARGUMENT;
    }
    const OOC_MapVtable* vtable = ooc_getInterfaceVtable(self, ooc_mapClass());
    if (!vtable || !vtable->clear) {
        return OOC_ERROR_NOT_IMPLEMENTED;
    }
    return vtable->clear(self);
}

void* ooc_mapKeySet(void* self) {
    if (!self) {
        return NULL;
    }
    const OOC_MapVtable* vtable = ooc_getInterfaceVtable(self, ooc_mapClass());
    if (!vtable || !vtable->keySet) {
        return NULL;
    }
    return vtable->keySet(self);
}

void* ooc_mapValues(void* self) {
    if (!self) {
        return NULL;
    }
    const OOC_MapVtable* vtable = ooc_getInterfaceVtable(self, ooc_mapClass());
    if (!vtable || !vtable->values) {
        return NULL;
    }
    return vtable->values(self);
}

File: src/interface/map/oocMap.r
#ifndef OOC_MAP_R_
#define OOC_MAP_R_

#include "oocInterface.r"
#include "oocError.h"
#include <stddef.h>
#include <stdbool.h>

typedef struct OOC_MapVtable OOC_MapVtable;
typedef struct OOC_MapClass OOC_MapClass;

struct OOC_MapVtable {
    size_t (*size)(void* self);
    bool (*isEmpty)(void* self);
    bool (*containsKey)(void* self, void* key);
    bool (*containsValue)(void* self, void* value);
    void* (*get)(void* self, void* key);
    OOC_Error (*put)(void* self, void* key, void* value);
    OOC_Error (*remove)(void* self, void* key);
    OOC_Error (*clear)(void* self);
    void* (*keySet)(void* self);
    void* (*values)(void* self);
};

struct OOC_MapClass {
    OOC_InterfaceClass interfaceClass;
    OOC_MapVtable vtable;
};

#define OOC_MAP_METHOD_SIZE             offsetof(OOC_MapClass, vtable.size)
#define OOC_MAP_METHOD_IS_EMPTY         offsetof(OOC_MapClass, vtable.isEmpty)
#define OOC_MAP_METHOD_CONTAINS_KEY     offsetof(OOC_MapClass, vtable.containsKey)
#define OOC_MAP_METHOD_CONTAINS_VALUE   offsetof(OOC_MapClass, vtable.containsValue)
#define OOC_MAP_METHOD_GET              offsetof(OOC_MapClass, vtable.get)
#define OOC_MAP_METHOD_PUT              offsetof(OOC_MapClass, vtable.put)
#define OOC_MAP_METHOD_REMOVE           offsetof(OOC_MapClass, vtable.remove)
#define OOC_MAP_METHOD_CLEAR            offsetof(OOC_MapClass, vtable.clear)
#define OOC_MAP_METHOD_KEY_SET          offsetof(OOC_MapClass, vtable.keySet)
#define OOC_MAP_METHOD_VALUES           offsetof(OOC_MapClass, vtable.values)

#endif
